### PR TITLE
Refactor completer plugin into a standalone plugin.

### PR DESCRIPTION
**Pull request by afshin**

### CODE REVIEW

File: src/common/instancetracker.ts [hunk 1]
```diff
@@ -224,6 +231,25 @@ class InstanceTracker<T extends Widget> implements IInstanceTracker<T>, IDisposa
   }
 
   /**
+   * Get a widget with a specified ID that is held by the tracker.
+   *
+   * @param - id The ID string of the widget.
```
Comment (blink1073) [inline src/common/instancetracker.ts]: perhaps "the id property"?

File: src/completer/plugin.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    };
+
+    app.commands.addCommand(CommandIDs.invoke, {
+      execute: args => {
+        let id = args['id'] as string;
+        if (!id) {
+          return;
+        }
+
+        const handler =handlers[id];
```
Comment (blink1073) [inline src/completer/plugin.ts]: missing space

File: src/completer/plugin.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+      // Listen for kernel changes.
+      panel.console.session.kernelChanged.connect((sender, kernel) => {
+        handler.kernel = kernel;
+      });
+    });
+
+    // Add console completer command.
+    app.commands.addCommand(CommandIDs.invokeConsole, {
+      execute: () => {
+        const id = consoles.currentWidget && consoles.currentWidget.id;
```
Comment (blink1073) [inline src/completer/plugin.ts]: bail if null?

File: src/inspector/plugin.ts [hunk 4]
```diff
@@ -113,37 +109,49 @@ const service: JupyterLabPlugin<IInspector> = {
...
 const consolePlugin: JupyterLabPlugin<void> = {
   id: 'jupyter.extensions.console-inspector',
-  requires: [IConsoleTracker],
-  provides: IInspector,
+  requires: [IInspector, IConsoleTracker],
   autoStart: true,
-  activate: (app: JupyterLab, consoles: IConsoleTracker): void => {
+  activate: (app: JupyterLab, manager: IInspector, consoles: IConsoleTracker): void => {
+    // Maintain association of new consoles with their respective handlers.
+    const handlers = new AttachedProperty<Widget, InspectionHandler>({
```
Comment (blink1073) [inline src/inspector/plugin.ts]: can use a simple `Map` directly and remove entries from the map when the widget is disposed.

File: src/inspector/plugin.ts [hunk 5]
```diff
@@ -156,36 +164,48 @@ const consolePlugin: JupyterLabPlugin<void> = {
...
 const notebookPlugin: JupyterLabPlugin<void> = {
   id: 'jupyter.extensions.notebook-inspector',
-  requires: [INotebookTracker],
-  provides: IInspector,
+  requires: [IInspector, INotebookTracker],
   autoStart: true,
-  activate: (app: JupyterLab, notebooks: INotebookTracker): void => {
+  activate: (app: JupyterLab, manager: IInspector, notebooks: INotebookTracker): void => {
+    // Maintain association of new notebooks with their respective handlers.
+    const handlers = new AttachedProperty<Widget, InspectionHandler>({
```
Comment (blink1073) [inline src/inspector/plugin.ts]: idem

File: src/tooltip/plugin.ts [hunk 6]
```diff
@@ -18,96 +18,193 @@ import {
...
+    // Add tooltip launch command.
+    app.commands.addCommand(CommandIDs.launchConsole, {
+      execute: () => {
+        const parent = consoles.currentWidget;
+
+        if (!parent) {
+          return;
+        }
 
+        const anchor = parent.console;
```
Comment (blink1073) [inline src/tooltip/plugin.ts]: store `console` and use it throughout here?

File: src/tooltip/plugin.ts [hunk 7]
```diff
@@ -18,96 +18,193 @@ import {
...
-      Widget.attach(tooltip, document.body);
 
-      // Make sure the parent notebook/console still has the focus.
-      parent.activate();
+/**
+ * A namespace for private data.
+ */
+namespace Private {
+  /**
+   * A counter for outstanding requests.
```
Comment (blink1073) [inline src/tooltip/plugin.ts]: gotta love those outstanding requests.

Comment (afshin) [inline src/tooltip/plugin.ts]: They're the best.

**Comment by blink1073** [comment]
noice

**Pull request merged**